types/address/hash.go

```diff
@@ -0,0 +1,73 @@
+package address
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"reflect"
+	"sort"
+	"unsafe"
+
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Len is the length of base addresses
+const Len = sha256.Size
+
+type Addressable interface {
+	Address() []byte
+}
+
+// Hash creates a new address from address type and key
+func Hash(typ string, key []byte) []byte {
```

Comment:
I wanted to use `sdk.Address` type here, but this causes import cycle. This is because `/types` uses `address.MaxAddressLen` const.
If we agree to use `sdk.Address` type here, then I propose:
* move all other address related functions to `/types/address` (preferred, more work) - preferably after pubkey-bech32 removal.
* move `MaxAddressLen` to `/types`.
*